### ISSUE TITLE
FR: Async custom cert verification API.

### ISSUE THREAD

**Issue by matthewstevenson88** [openssl/openssl#14945]
I'm interested in having an API that allows an application to configure the `SSL_CTX` to do async custom cert verification since, IIUC, the `SSL_CTX_set_cert_verify_callback` method is only used for sync custom cert verification. I'm hoping this is technically possible to do so, since BoringSSL added this support a few years ago. This API is needed by [gRPC](https://github.com/grpc) for a custom cert verification feature that we'd like to have.

To drive the conversation forward, I had a few questions:

1. Is such an API already planned?
2. If not, are you open to having such an API in OpenSSL?
3. If you're open to this idea, do you have a conservative estimate for how long it might take for this API to appear in a release? Would this estimate change if we contributed some/all of the development work?

Thanks for your help! :)

cc: @davidben

**Comment by mattcaswell** [comment]
For client side verification of the server certificate this support was recently added to the master branch (OpenSSL 3.0) in #13906. There's currently no plans to do it for server side verification of the client certificate.

**Comment by matthewstevenson88** [comment]
Thanks @mattcaswell! Would you be open to having analogous functionality for the server side?

Ignoring for now the issue of prioritizing the work or who would implement it. I'm curious as to whether there is opposition to having the server side functionality on principle.

**Comment by mattcaswell** [comment]
> Would you be open to having analogous functionality for the server side?

Yes.

> Ignoring for now the issue of prioritizing the work or who would implement it. I'm curious as to whether there is opposition to having the server side functionality on principle.

From a priority perspective the dev team are fully focused on getting OpenSSL 3.0 completed. We are very close to feature freeze and we are at a point where we are throwing out stuff which we don't need to keep the scope under control. Most likely a PR adding this would not be merged prior to 3.0 completion later this year, i.e. it probably wouldn't be included in a release before 3.1. Even then its likely that the PR would have to be submitted from a someone external. Such a feature is probably a "nice to have" rather than a priority item.

**Comment by matthewstevenson88** [comment]
Thanks for clarifying! Could you give a rough estimate for when 3.0 and 3.1 are each expected to be released? I saw an old timeline for 3.0 on the wiki but it seems a bit out-of-date.

I'll circle back to this once 3.0 is out to see if we can get it in 3.1.

**Comment by mattcaswell** [comment]
For 3.0 final we are hoping for something roughly in the September timeframe. No time line yet for 3.1.

**Comment by nhorman** [comment]
@matthewstevenson88 did you file a PR for this?

**Comment by matthewstevenson88** [comment]
@nhorman No, I have not put together a PR for this. For my use case, we were able to find a workaround. Leaving this issue open though since there seem to be other users requesting this too.

**Comment by nhorman** [comment]
@uncbailey1 since you expressed interest int this, same question to you, are you interested in writing a PR for this, given that the core team doesn't currently have plans for this?

**Comment by nhorman** [comment]
no response, marking as inactive

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
